Pseudo-code

dnxjay opened the pull request:
Add linear regression fitting function with weighted error handling and docstrings

dnxjay committed:
Add linear regression fitting function with error handling and docstrings
testing to be done

dnxjay commented:
@kylemasc917 would it be best to have @SchrodingersStruggle review this?

SchrodingersStruggle commented:
@dnxjay @kylemasc917 Would it make sense for me to review this, or rather the person who will be responsible for creating unit tests for it? I would certainly be willing to though, especially if nobody has decided to claim the unit testing yet. I still will be doing the least square fitting though.

dnxjay commented:
> @dnxjay @kylemasc917 Would it make sense for me to review this, or rather the person who will be responsible for creating unit tests for it? I would certainly be willing to though, especially if nobody has decided to claim the unit testing yet. I still will be doing the least square fitting though.

That's a good point. i also don't mind doing the test as well, maybe @kylemasc917 can give some input on this.

SchrodingersStruggle commented:
I actually just noticed I was given the task with you of translating the pseudo code, when what I volunteered for in #1 was actually producing the least square fit, so I am actually a bit confused on the plans. Am I actually assigned with the unit testing? Hopefully I can get this figured out soon.

dnxjay commented:
> I actually just noticed I was given the task with you of translating the pseudo code, when what I volunteered for in #1 was actually producing the least square fit, so I am actually a bit confused on the plans. Am I actually assigned with the unit testing? Hopefully I can get this figured out soon.

im pretty sure i was the one given this task during lecture but i could be wrong

laserlab commented:
Yes so @SchrodingersStruggle is assigned #9 now since I got the generating part misunderstood. @haiderabbas007 is assigned now to this task for documentation and unit testing

laserlab commented:
so it would make sense to check if @haiderabbas007 wants to review it, but it’s time critical so it gets merged before class tomorrow.

kylemasc917 commented:
I added both @SchrodingersStruggle and @haiderabbas007 please both review it but if by class on Monday only one has I will still merge it.

laserlab commented:
So after discussing with @SchrodingersStruggle it makes most sense to switch tasks so it gets reviewed fast

dnxjay commented:
> So after discussing with @SchrodingersStruggle it makes most sense to switch tasks so it gets reviewed fast
Okay sweet, once i receive the review ill make the necessary changes if needed so we can get this merged asap

SchrodingersStruggle commented on pseudo_code/linear_fit.py:
```diff
@@ -0,0 +1,54 @@
+    a = (s_y - s_x * b) / S
+    
+    # Calculate error estimates for a and b
+    sigma_a2 = (1 + s_x**2 / (S * s_tt)) / S
```
I think it may be more beneficial to return the standard deviation of the slope and intercept, rather than variance. This is a much more common form of establishing error values, and this is done simply by taking the square root of this variance calculated.

SchrodingersStruggle commented on pseudo_code/linear_fit.py:
```diff
@@ -0,0 +1,54 @@
+def linear_fit(data):
```
What type of data is expected in this input? This is not a request for a change, more elaboration, since if it is expected to be a csv file, the code will need to be adjusted a bit.

SchrodingersStruggle replied:
I honestly think it may be easier to make the parameters for input in this function be x, y, sigma as lists of each, just to make it easier to work with. If you want to have the input be a csv file, this requires a bit more touchup such as using pandas to read the csv and extract values out of it.
This would require the csv the data is found in to have been read by a different function before using this function, so idk. Maybe @kylemasc917 can give his advice.
```suggestion
def linear_fit(x, y, sigma):
```

SchrodingersStruggle reviewed (commented):
I added some things to review for discussion. Mostly the input parameter's type, and adjusting to output error as standard deviation rather than variance. If you agree to change the input parameters, a couple of changes in the module need to be made, and I can help with suggestions for that if need be.

dnxjay committed:
Made some adjustments based on the reviewers comments which were accurate. Refactor linear_fit function for modularity and improved usability
- Updated linear_fit to accept x, y, and sigma as separate lists instead of tuples
- Added read_data_from_csv function to read x, y, and sigma values from a CSV file using pandas
- Changed variance calculations for slope and intercept to return standard deviations for clearer error representation
- Improved error handling and code readability based on reviewer feedback

dnxjay commented:
@SchrodingersStruggle i made the necessary adjustments. The points you brought up were valid. Hopefully this new code fixes the concerns

SchrodingersStruggle reviewed (approved):
Looks good. I see you added some docstrings which was meant to be part of my tasks, but that's okay. I will just add the initial docstring, and maybe edit them a bit when my time comes. A few formatting tweaks can be made, but I won't nitpick for the sake of being merged before lecture, and rather I will address it with my own pull request later on.

I tested it in vscode using a mock csv I created and it holds up nicely.

HeshamElsaman commented:
Sounds good, merging it right now.
